xformers/components/attention/csrc/cuda/mem_eff_attention/kernel_forward.h

```diff
@@ -628,6 +729,67 @@ struct AttentionKernel {
 
       __syncthreads();
 
+      // apply dropout (if applicable) after we've written Pij to smem.
+      // dropout is applied by multiplying each element of Pij by:
+      // - 0 with probability dropout_p
+      // - 1 / (1 - dropout_p) with probability 1 - dropout_p
+      //
+      // for backward purposes we want to be able to map each element of the attention
+      // matrix to the same random uniform number as the one we used in forward,
+      // without needing to use the same iteration order or having to store the dropout
+      // matrix.
+      // its possible to do this in registers but it ends up being very slow because
+      // each thread having noncontiguous strips of the Pij tile means
+      // we have to skip around a lot, and also have to generate a single random
+      // number at a time
+      if (p.use_dropout) {
+        auto si = shared_storage.after_mm0.si.accum_ref();
+        // each thread handles a contiguous sequence of elements from Sij, all coming
+        // from the same row.
+        // the reason they have to come from the same row is that the sampling random
+        // numbers from a contiguous random number sequence is much more efficient
+        // than jumping around, and the linear offset of each element of
+        // S (the global matrix) maps to an offset in a random number sequence.
+        // for S, the end of a row and the beginning of the next have adjacent offsets,
+        // but for Sij, this is not necessarily the case.
+        const int num_threads = blockDim.x * blockDim.y * blockDim.z;
+        const int threads_per_row = cutlass::fast_min(
+            num_threads / problem_size_0_m, problem_size_0_n);
+        const int elts_per_thread = cutlass::round_nearest(
+            cutlass::ceil_div(problem_size_0_n, threads_per_row), 4);
+
+        const int thread_i = thread_id() / threads_per_row;
+        const int thread_start_j = (thread_id() % threads_per_row) * elts_per_thread;
+
+        if (thread_i < problem_size_0_m && thread_start_j < problem_size_0_n) {
+          curandStatePhilox4_32_10_t curand_state = curand_state_init;
+          skipahead(
+              (query_start + thread_i) * p.num_keys + (iter_key_start + thread_start_j),
+              &curand_state);
```

Comment:
This does not build [on CI](https://app.circleci.com/pipelines/github/facebookresearch/xformers/4086/workflows/65c784e9-4a28-4ffb-91e4-4531b59a1ee9/jobs/13911/steps) (maybe because it's cuda 11.4?):
```bash
    /usr/local/cuda-11.4/include/curand_kernel.h(1255): error: class "curandStatePhilox4_32_10" has no member "x"
              detected during:
                instantiation of "void skipahead(unsigned int, T) [with T=curandStatePhilox4_32_10_t *]"
    /home/circleci/xformers/xformers/components/attention/csrc/cuda/mem_eff_attention/kernels/../kernel_forward.h(768): here
                instantiation of "void AttentionKernel<scalar_t_, ArchTag, isAligned_, kQueriesPerBlock, kKeysPerBlock, kSingleValueIteration>::attention_kernel(AttentionKernel<scalar_t_, ArchTag, isAligned_, kQueriesPerBlock, kKeysPerBlock, kSingleValueIteration>::Params &) [with scalar_t_=cutlass::bfloat16_t, ArchTag=cutlass::arch::Sm50, isAligned_=false, kQueriesPerBlock=32, kKeysPerBlock=128, kSingleValueIteration=true]"
    /home/circleci/xformers/xformers/components/attention/csrc/cuda/mem_eff_attention/kernels/forward_bf16.cu(4): here
```

Reply:
looking

Reply:
fixed